[Future] modularized running callbacks on future completion

osalpekar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#39114 [Future] modularized running callbacks on future completion**

`setErrorInternal` and `markCompletedInternal` have the exact same code to run callbacks on completion (move callbacks to stack vector, release lock, notify waiting threads, then run callbacks in a loop). This PR introduces a new method with this functionality to dedupe code from the above 2 functions.

[Open to thoughts/feedback about whether this is a good idea]

Differential Revision: [D21753987](https://our.internmc.facebook.com/intern/diff/D21753987/)

osalpekar committed:
[Future] modularized running callbacks on future completion

`setErrorInternal` and `markCompletedInternal` have the exact same code to run callbacks on completion (move callbacks to stack vector, release lock, notify waiting threads, then run callbacks in a loop). This PR introduces a new method with this functionality to dedupe code from the above 2 functions.

[Open to thoughts/feedback about whether this is a good idea]

Differential Revision: [D21753987](https://our.internmc.facebook.com/intern/diff/D21753987/)

[ghstack-poisoned]

osalpekar committed:
Update on "[Future] modularized running callbacks on future completion"


`setErrorInternal` and `markCompletedInternal` have the exact same code to run callbacks on completion (move callbacks to stack vector, release lock, notify waiting threads, then run callbacks in a loop). This PR introduces a new method with this functionality to dedupe code from the above 2 functions.

[Open to thoughts/feedback about whether this is a good idea]

Differential Revision: [D21753987](https://our.internmc.facebook.com/intern/diff/D21753987/)

[ghstack-poisoned]

osalpekar committed:
Update on "[Future] modularized running callbacks on future completion"


`setErrorInternal` and `markCompletedInternal` have the exact same code to run callbacks on completion (move callbacks to stack vector, release lock, notify waiting threads, then run callbacks in a loop). This PR introduces a new method with this functionality to dedupe code from the above 2 functions.

[Open to thoughts/feedback about whether this is a good idea]

Differential Revision: [D21753987](https://our.internmc.facebook.com/intern/diff/D21753987/)

[ghstack-poisoned]

rohan-varma commented on torch/csrc/utils/future.h:
```diff
@@ -163,13 +153,19 @@ class TORCH_API Future final {
+  }
+
+  // Runs all the callbacks attached to this future.
+  void executeCallbacksOnCompletion(std::unique_lock<std::mutex>& lock) {
```
I think passing a ref to this lock around functions (this existed before this PR as well) could potentially get confusing, and it would be the developer's responsibility to unlock it properly. What about deduping the code into 2 portions:

1: setting error/value and completed, and moving the callbacks into a local vector
(unlock between these)
2: notifying on the CV and running end callbacks

osalpekar committed:
Update on "[Future] modularized running callbacks on future completion"


`setErrorInternal` and `markCompletedInternal` have the exact same code to run callbacks on completion (move callbacks to stack vector, release lock, notify waiting threads, then run callbacks in a loop). This PR introduces a new method with this functionality to dedupe code from the above 2 functions.

[Open to thoughts/feedback about whether this is a good idea]

Differential Revision: [D21753987](https://our.internmc.facebook.com/intern/diff/D21753987/)

[ghstack-poisoned]

osalpekar committed:
Update on "[Future] modularized running callbacks on future completion"


`setErrorInternal` and `markCompletedInternal` have the exact same code to run callbacks on completion (move callbacks to stack vector, release lock, notify waiting threads, then run callbacks in a loop). This PR introduces a new method with this functionality to dedupe code from the above 2 functions.

[Open to thoughts/feedback about whether this is a good idea]

Differential Revision: [D21753987](https://our.internmc.facebook.com/intern/diff/D21753987/)

[ghstack-poisoned]
